### PR TITLE
Framework: Extract "edit-post" to its own module

### PR DESCRIPTION
**Pull request by youknowriad**
Sorry for the huge PR, sometimes there's no alternative.

This PR extracts the `edit-post` module to its own module. This surfaced some "issues"

 - The need to spit the store into two stores. One store for the "editor" module (library) which is more a "local state" to the `EditorProvider` component. This store is saved in the React context using a different key "editorStore" to avoid ambiguity when calling `connect`

 - Some components in `edit-post` where still needed access to the `editor` module's store, in general this is fixed by adding reusable components to the `editor` module.

 - The "Show block inspector" button proved to be a challenge, because the idea is that clicking this button shows the inspector, but since the BlockInspector is a reusable component that can be shown anyware in the application using the "editor" module, we needed a `showBlockInspector` to allow the surrouding application (`edit-post`) to handle this button properly.

I'd appreciate quick feedback and opinions, this won't be enought to keep without conflicts and continuous rebasing.

**Testing instructions**

 - This may introduce some unnoticed regressions, heavily testing this PR is necessary.

### CODE REVIEW

**Comment by gziolo** [comment]
I took a quick glance at this code. Looks pretty good.

> The "Show block inspector" button proved to be a challenge, because the idea is that clicking this button shows the inspector, but since the BlockInspector is a reusable component that can be shown anyware in the application using the "editor" module, we needed a `showBlockInspector` to allow the surrouding application (`edit-post`) to handle this button properly.

I'm not so happy about passing it down to all those components. I will think about some alternatives.

File: editor/constants.js [hunk 1]
```diff
@@ -1,7 +1,7 @@
 /**
  * Internal dependencies
  */
-import scssVariables from '!!sass-variables-loader!./assets/stylesheets/_variables.scss';
+import scssVariables from '!!sass-variables-loader!../edit-post/assets/stylesheets/_variables.scss';
```
Comment (gziolo) [inline editor/constants.js]: It looks like some SCSS variables should remain locate in `editor/` folder.

Comment (youknowriad) [reply editor/constants.js]: Yeah, we have this issue for all components, these variables and mixins are not modularized properly. They are used in all the modules "components, "block", "editor", "edit-post" and a lot of them are specific to a module.

Comment (gziolo) [reply editor/constants.js]: We should tackle this separately in that case.

**Comment by gziolo** [comment]
We discussed it extensively on Slack, but let me summarize my feedback here.

So as I understand, the idea here is to have isolated Redux state for the editor and a separate one for `edit-post` or any other type that we introduce in the future. It's not ideal because all contributors will have to pass the store name explicitly in the `connect` method. We might want to create local `connect` method and/or document it properly. We could add lint option to make sure `editor` library uses `connect` method properly.

I’m not happy about `onShowInspector`, it is connected at the UI level with `Settings` button:

screen shot 2017-11-27 at 12 26 12

All other toolbar buttons are handled within the editor, this one is not. I guess one of the reasons is because we can’t expose its current state to the `edit-post` library. Another one is as @youknowriad and @mtias suggested is that some apps might want to display advanced setting in a modal. I think the latter is a good reason to pass down this callback to make it easy to customize. I still hope we can find a solution that will simplify exposing this method to the nested components.

I was exploring if we could reuse pattern from `<BlockCount />` to tackle `isSidebarOpened` flag differently. We could rename it to sth like `areAdvancedSettingActive` and still keep in the editor's state. We could expose it as `<AdvancedSettings />` component and use this way inside `edit-post` library:
```js
<AdvancedSettings>
        { ( areAdvancedSettingActive, closeAdvancedSettings ) => areAdvancedSettingActive && <Sidebar onClose={ closeAdvancedSettings } /> }
</AdvancedSettings>
```
The obvious issue here is that it won't play well with the modal case.

It would be nice to gather feedback from @mkaz and @georgeh how this proposal fits with their work on P2 integration. It would be a very valuable to hear from them before we proceed.

**Comment by gziolo** [comment]
One more thing to consider. Previously only editor library was using Redux libraries. This PR changes that, which will cause both editor and edit-post to bundle the same files twice. We should either make those libs external or put them inside shared wp.store library similat to what wp.element does.

**Comment by youknowriad** [comment]
I'm closing this right now, I'll do another attempt once the "data" module finalized, which will help with state access across modules.

**Pull request closed without merging**